src/modules/emote_autocomplete/twitch/EmoteAutocomplete.jsx

```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {getCurrentUser} from '../../../utils/user.js';
+import EmoteWhisper from '../components/EmoteWhisper.jsx';
+import settings from '../../../settings.js';
+import {SettingIds} from '../../../constants.js';
+import twitch, {SelectionTypes} from '../../../utils/twitch.js';
+
+let mountedNode;
+
+const CHAT_TEXT_AREA = 'textarea[data-a-target="chat-input"], div[data-a-target="chat-input"]';
+const EMOTE_AUTOCOMPLETE_CONTAINER_SELECTOR = 'div[data-a-target="bttv-autocomplete-matches-container"]';
+
+export default class EmoteAutocomplete {
+  constructor() {
+    this.load();
+    settings.on(`changed.${SettingIds.EMOTE_AUTOCOMPLETE}`, () => this.load());
+  }
+
+  load() {
+    if (getCurrentUser() == null) {
+      return;
+    }
+
+    const emoteAutcompleteContainer = document.querySelector(EMOTE_AUTOCOMPLETE_CONTAINER_SELECTOR);
+    const emoteAutocomplete = settings.get(SettingIds.EMOTE_AUTOCOMPLETE);
+
+    if (emoteAutcompleteContainer == null && emoteAutocomplete) {
+      const element = document.querySelector(CHAT_TEXT_AREA);
+      const whisperContainer = document.createElement('div');
+      whisperContainer.setAttribute('data-a-target', 'bttv-autocomplete-matches-container');
+      document.body.appendChild(whisperContainer);
+
+      if (mountedNode != null) {
+        ReactDOM.unmountComponentAtNode(mountedNode);
+      }
+
+      ReactDOM.render(
+        <EmoteWhisper
+          boundingQuerySelector={CHAT_TEXT_AREA}
+          chatInputElement={element}
+          onComplete={this.replaceChatInputPartialEmote}
+          getChatInputPartialEmote={this.getChatInputPartialEmote}
+        />,
+        whisperContainer
+      );
+
+      mountedNode = whisperContainer;
+    }
+
+    if (!emoteAutocomplete && emoteAutcompleteContainer != null) {
+      ReactDOM.unmountComponentAtNode(mountedNode);
+    }
+  }
+
+  isActive() {
+    return settings.get(SettingIds.EMOTE_AUTOCOMPLETE) && this.getChatInputPartialEmote() != null;
+  }
+
+  replaceChatInputPartialEmote({code}) {
+    const currentValue = twitch.getChatInputValue();
+
+    const lastSpaceIndex = currentValue.lastIndexOf(' ');
+    const newValue = lastSpaceIndex < 0 ? code : `${currentValue.substring(0, lastSpaceIndex)} ${code}`;
```

Comment:
im curious: why not use the same regex for getting the input, and replacing the input?

```
const PARTIAL_EMOTE_REGEX = /:([a-z0-9-_.:]+?)(:|$)/i;
...
retest.exec(':test:') -> [':test:', 'test', ':', index: 0, input: ':test:', groups: undefined]
'test :testa:'.replace(retest, 'lol') -> 'test lol'
```